src/Data/List/Types.purs

```diff
@@ -0,0 +1,167 @@
+module Data.List.Types where
+
+import Prelude
+
+import Control.Alt (class Alt)
+import Control.Alternative (class Alternative)
+import Control.Comonad (class Comonad)
+import Control.Extend (class Extend)
+import Control.MonadPlus (class MonadPlus)
+import Control.MonadZero (class MonadZero)
+import Control.Plus (class Plus)
+
+import Data.Foldable (class Foldable, foldr, foldl, intercalate)
+import Data.Generic (class Generic)
+import Data.Maybe (Maybe(..))
+import Data.Monoid (class Monoid, mempty)
+import Data.Newtype (class Newtype)
+import Data.NonEmpty (NonEmpty, (:|))
+import Data.NonEmpty as NE
+import Data.Traversable (class Traversable, traverse, sequence)
+import Data.Tuple (Tuple(..))
+import Data.Unfoldable (class Unfoldable)
+
+data List a = Nil | Cons a (List a)
+
+infixr 6 Cons as :
+
+derive instance genericList :: Generic a => Generic (List a)
+
+instance showList :: Show a => Show (List a) where
+  show Nil = "Nil"
+  show xs = "(" <> intercalate " : " (show <$> xs) <> " : Nil)"
+
+instance eqList :: Eq a => Eq (List a) where
+  eq xs ys = go xs ys true
+    where
+      go _ _ false = false
+      go Nil Nil acc = acc
+      go (x : xs) (y : ys) acc = go xs ys $ acc && (y == x)
+      go _ _ _ = false
+
+instance ordList :: Ord a => Ord (List a) where
+  compare xs ys = go xs ys
+    where
+    go Nil Nil = EQ
+    go Nil _ = LT
+    go _ Nil = GT
+    go (x : xs) (y : ys) =
+      case compare x y of
+        EQ -> go xs ys
+        other -> other
+
+instance semigroupList :: Semigroup (List a) where
+  append Nil ys = ys
+  append (x : xs) ys = x : (xs <> ys)
+
+instance monoidList :: Monoid (List a) where
+  mempty = Nil
+
+instance functorList :: Functor List where
+  map f = foldr (\x acc -> f x : acc) Nil
+
+instance foldableList :: Foldable List where
+  foldr f b as = foldl (\g a x -> g (f a x)) id as b
```

Comment:
If you are folding with continuations, how is this stack safe? The folding may be stack safe, but running the composition of continuations will blow the stack won't it (there is no test for the stack safety of foldr)?